hir/src/asm.rs

```diff
@@ -0,0 +1,2447 @@
+        assert!(
+            offset < 4,
+            "invalid element offset, must be in the range 0..=3, got {}",
+            offset
+        );
+        self.stack.drop();
+        self.asm
+            .push(self.ip, MasmOp::MemLoadOffsetImm(addr, offset));
+    }
+
+    /// Pops an element containing a memory address from the top of the stack,
+    /// and loads the word at that address to the top of the stack.
+    pub fn loadw(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemLoadw);
+    }
+
+    /// Loads the word at the given address to the top of the stack.
+    pub fn loadw_imm(self, addr: u32) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemLoadwImm(addr));
+    }
+
+    /// Pops two elements, the first containing a memory address from the top of the stack,
+    /// the second the value to be stored as the first element of the word at that address.
+    pub fn store(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStore);
+    }
+
+    /// Pops an element from the top of the stack, and stores it as the first element of
+    /// the word at the given address.
+    pub fn store_imm(self, addr: u32) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStoreImm(addr));
+    }
+
+    /// Pops two elements, the first containing a memory address + element offset from the
+    /// top of the stack, the second the value to be stored to the word at that address,
+    /// using the offset to determine which element will be written to.
+    pub fn store_offset(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStoreOffset);
+    }
+
+    /// Pops an element from the top of the stack, and stores it at the given offset of
+    /// the word at the given address.
+    pub fn store_offset_imm(self, addr: u32, offset: u8) {
+        assert!(
+            offset < 4,
+            "invalid element offset, must be in the range 0..=3, got {}",
+            offset
+        );
+        self.stack.drop();
+        self.asm
+            .push(self.ip, MasmOp::MemStoreOffsetImm(addr, offset));
+    }
+
+    /// Pops an element containing a memory address from the top of the stack,
+    /// and then pops a word from the stack and stores it as the word at that address.
+    pub fn storew(self) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStorew);
+    }
+
+    /// Pops a word from the stack and stores it as the word at the given address.
+    pub fn storew_imm(self, addr: u32) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::MemStorewImm(addr));
+    }
+
+    /// Pops a boolean value from the stack, and executes the first block if it is true,
+    /// otherwise the second block.
+    pub fn if_true(self, then_blk: MasmBlockId, else_blk: MasmBlockId) {
+        self.stack.drop();
+        self.asm.push(self.ip, MasmOp::If(then_blk, else_blk))
```

Comment:
Similar comment to the one for `while_true`: How do you keep track of the stack height at the join point between the branches?

Reply:
Same deal as with `while_true`, with the same general approach to addressing it. Tthis is more generally handled as part of stackification, but when it comes to constructing inline assembly, we probably want to add some additional validation to catch the mistake earlier when defining the assembly, rather than during codegen, just like with the loops.